config-model/src/main/java/com/yahoo/vespa/model/application/validation/BundleValidator.java

```diff
@@ -117,27 +133,91 @@ private static void validateImportedPackages(DeployLogger deployLogger, String f
         });
     }
 
-    private enum DeprecatedArtifact {
+    private static final Pattern POM_FILE_LOCATION = Pattern.compile("META-INF/maven/.+?/.+?/pom.xml");
+
+    private Optional<String> getPomXmlContent(DeployLogger deployLogger, JarFile jarFile) {
+        return jarFile.stream()
+                .filter(f -> POM_FILE_LOCATION.matcher(f.getName()).matches())
+                .findFirst()
+                .map(f -> {
+                    try {
+                        return new String(jarFile.getInputStream(f).readAllBytes());
+                    } catch (IOException e) {
+                        deployLogger.log(Level.INFO,
+                                String.format("Unable to read '%s' from '%s'", f.getName(), jarFile.getName()));
+                        return null;
+                    }
+                });
+    }
+
+    private void validatePomXml(DeployLogger deployLogger, String jarFilename, String pomXmlContent) {
+        try {
+            Document pom = DocumentBuilderFactory.newDefaultInstance().newDocumentBuilder()
+                    .parse(new InputSource(new StringReader(pomXmlContent)));
+            NodeList dependencies = (NodeList) XPathFactory.newDefaultInstance().newXPath()
+                    .compile("/project/dependencies/dependency")
+                    .evaluate(pom, XPathConstants.NODESET);
+            for (int i = 0; i < dependencies.getLength(); i++) {
+                Element dependency = (Element) dependencies.item(i);
+                String groupId = dependency.getElementsByTagName("groupId").item(0).getTextContent();
+                String artifactId = dependency.getElementsByTagName("artifactId").item(0).getTextContent();
+                for (DeprecatedMavenArtifact deprecatedArtifact : DeprecatedMavenArtifact.values()) {
+                    if (groupId.equals(deprecatedArtifact.groupId) && artifactId.equals(deprecatedArtifact.artifactId)) {
+                        deployLogger.logApplicationPackage(Level.WARNING,
+                                String.format("For pom.xml in '%s': \n" +
+                                                "The dependency %s:%s is listed below dependencies. \n" +
+                                                "%s",
+                                        jarFilename, groupId, artifactId, deprecatedArtifact.description));
+                    }
+                }
+            }
+        } catch (ParserConfigurationException e) {
+            throw new RuntimeException(e);
+        } catch (Exception e) {
+            deployLogger.log(Level.INFO, String.format("Unable to parse pom.xml from %s", jarFilename));
+        }
+    }
+
+    private enum DeprecatedMavenArtifact {
+        VESPA_HTTP_CLIENT_EXTENSION("com.yahoo.vespa", "vespa-http-client-extensions",
+                "The 'vespa-http-client-extensions' artifact will be removed in Vespa 8. " +
```

Comment:
It seems unnecessary to repeat the artifact name? Couldn't we just say "This artifact"?

Reply:
Fixed